src/Features/CSharp/Portable/Diagnostics/Analyzers/FileBasedPrograms/FileLevelDirectiveDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.CodeAnalysis.CodeQuality;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.DotNet.FileBasedPrograms;
+
+namespace Microsoft.CodeAnalysis.FileBasedPrograms;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+internal sealed class FileLevelDirectiveDiagnosticAnalyzer()
+    : AbstractCodeQualityDiagnosticAnalyzer(
+        descriptors: [Rule],
+        generatedCodeAnalysisFlags: GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics)
+{
+    public const string DiagnosticId = "FileBasedPrograms";
+
+#pragma warning disable RS0030 // Do not use banned APIs
+    // We would use 'AbstractCodeQualityDiagnosticAnalyzer.CreateDescriptor()', but, we want these diagnostics to have error severity.
+    private static readonly DiagnosticDescriptor Rule = new(
+                id: DiagnosticId,
+                title: DiagnosticId,
+                messageFormat: "{0}",
+                category: "Usage",
+                defaultSeverity: DiagnosticSeverity.Error,
+                isEnabledByDefault: true,
+                helpLinkUri: "https://learn.microsoft.com/dotnet/csharp/language-reference/preprocessor-directives#file-based-apps",
+                // Note that this is an "editor-only" analyzer.
+                // When building or running file-based apps, the dotnet cli uses its own process to report errors on file-level directives.
+                customTags: DiagnosticCustomTags.Create(isUnnecessary: false, isConfigurable: false, isCustomConfigurable: false, enforceOnBuild: EnforceOnBuild.Never));
+#pragma warning restore RS0030 // Do not use banned APIs
+
+    public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+        => DiagnosticAnalyzerCategory.SyntaxTreeWithoutSemanticsAnalysis;
+
+    protected override void InitializeWorker(AnalysisContext context)
+    {
+        context.RegisterCompilationStartAction(context =>
+        {
+            context.RegisterSyntaxTreeAction(visitSyntaxTree);
+        });
+
+        void visitSyntaxTree(SyntaxTreeAnalysisContext context)
```

Comment:
ide uses PascalCase for local functinos.  in this case though, i would just inline.